### PR TITLE
fix(port-scanner): detect detached agent dev servers

### PR DESCRIPTION
**Pull request by Kitenite**
## What & why

Detect dev servers that Claude/Codex launch in detached subprocesses. Once a wrapper exits, these processes can be reparented to PID 1 with no controlling TTY, so walking descendants of the terminal shell misses their listening ports.

- Read the process table once per scan and retain existing tree-based attribution.
- Resolve remaining processes through inherited `SUPERSET_TERMINAL_ID` / `SUPERSET_PANE_ID`, propagating ownership from readable ancestors.
- Read environments via `ps -Eww` on macOS and procfs on Linux; re-read each scan because PID/parent alone cannot safely identify a process instance. Windows remains tree-only.
- Cover detached attribution, environment parsing, PID reuse, cancellation, and PortManager integration with regression tests.

- Revalidate current process ownership and the listening socket before killing a tracked port; do not rely on a possibly coalesced force-scan.
- Strip observed command arguments from macOS environment output before looking for terminal IDs.
- Reject results from replaced terminal sessions and preserve prior state when the environment snapshot fails. Inspection errors are sanitized so command stdout cannot leak environments into logs.

Scope: the owning terminal must remain registered. Workspace-level attribution after that terminal closes is not included.

## How I tested it

- `bun run lint`: passed.
- `bun run typecheck`: passed (38 tasks).
- Port-scanner tests: macOS 110 passed; Linux 109 passed, 1 macOS-only skip.
- `bun run check:plugins`: passed.
- `git diff --check`: passed.
- `bun run test`: incomplete; stopped in `@superset/trpc` with `Invalid environment variables` during module loading (required service credentials/configuration were undefined).

### Additional safety and edge-case verification

- Regression tests cover stale close requests after ownership changes, stopped listeners, terminal replacement during scanning/kill validation, failed inspection/recovery, and closing sibling ports on a shared process.
- Argument-only terminal IDs are rejected in parser tests, real targeted/whole-table macOS reads, and a real PPID-1 listener with cleared environment variables.
- Synthetic before/after checks against the previous commit: a stale entry invoked the mock kill function before (1 call), but not after (0 calls); an argv-only ID was accepted before and rejected after. No OS signal was sent by the mock-kill reproduction.
- Stress coverage: 10,002 synthetic process entries across three scans without mixing two terminal owners. Paired command/environment reads took 148–205 ms across 1,599 real macOS processes (10 samples, not a sustained battery/CPU soak).
- **Claude UI check:** a real Claude invocation in the matched app terminal launched the controlled double-fork/setsid listener (PPID 1, no TTY, port 18732). The sidebar showed one port; expanding showed localhost:18732, and the UI close confirmation stopped it. Screenshots were captured locally.
- **Codex partial check:** workspace-write initially blocked socket binding; enabling network access for the test invocation allowed the controlled listener to start. Its inherited terminal ID, PPID 1, socket and resolver attribution were verified. A code-triggered app restart returned the dev profile to onboarding, so this is **not** a Codex sidebar or app-restart restoration pass.
- Both agents ran a controlled self-detaching script, not the actual sprout-casquette dev-server/hot-reload journey. Test listeners and the dev stack started for verification were stopped.
- Final checks: macOS/Bun 1.3.11 tests 110 pass; Linux ARM64/Bun 1.3.14 tests 109 pass, 1 macOS-only skip; lint, typecheck and plugin validation pass. Root tests still stop at the previously documented missing tRPC environment variables.

**Remaining boundaries:** revalidation is not atomic with PID signalling or later process-tree escalation; no kernel process handle/start-time identity was added. macOS ps output remains a best-effort hint (unescaped environment values, non-atomic reads), not a security boundary. An unreadable orphan with no readable ancestor cannot be attributed. Windows runtime, sustained load/battery usage, and real dev-server hot reload through app restart remain unverified.

### Review follow-up (`bdca4d9317`)

- Removed the cross-scan environment cache to prevent stale ownership when a PID is reused with the same parent. Added resolver and PortManager regressions, including removal from the old terminal's kill controls after rescanning.
- `runTolerant` only tolerates numeric child exit codes; maxBuffer overflow and spawn failures now propagate. Real subprocess regression tests cover both.
- Linux environment reads receive the AbortSignal and propagate cancellation while preserving null results for ordinary read failures. Mocked filesystem tests cover this path on macOS.
- Environment reads measured 86–94 ms per scan across 1,427 processes on this Mac (five runs). This replaces near-zero cached steady-state work; batching and existing scan cadence/idle decay are unchanged.
- The checks above were rerun for this follow-up; the CDP screenshots below were captured for the original commit, not rerun for the review-only hardening.

### Before/after desktop verification

Used the authenticated dev Electron renderer from this worktree (`localhost:7145`, CDP `9555`, API `localhost:7141`). Through real CDP keyboard/mouse input, opened a workspace terminal and launched a controlled listener that double-forks and calls `setsid`. Both runs produced a listener on port 18732 with PPID 1, no controlling TTY, and an inherited terminal ID.

| Observation | Before fix | After fix |
| --- | --- | --- |
| Detached listener exists | Yes | Yes |
| Sidebar active-port badge | Absent | 1 active port |
| Expanded port details | No detected port | `localhost:18732` |
| Close port through UI | Not available | Listener killed; badge removed |

The app restarted between versions, so the same launch was repeated in a fresh terminal after the fix. Screenshots and numeric CDP measurements were captured locally in `.superset/cdp-evidence/` (not attached to this PR). The authenticated host-service query also confirmed the expected listener PID, terminal, and workspace.

This is end-to-end UI verification with a controlled detached-process launcher, not an actual Claude/Codex launch in the reported sprout-casquette workspace. The earlier scanner-level check against that workspace is separate evidence.

## Checklist

- [x] PR title follows conventional commits (`type(scope): subject`)
- [x] `bun run lint` and `bun run typecheck` pass (CI fails on lint warnings too)
- [ ] "Allow edits from maintainers" is checked on fork PRs — not applicable (same-repository branch)

---
## Summary by cubic
Fixes port detection for dev servers that Claude Code and Codex launch as detached background processes, so their listening ports now show up in the owning terminal's sidebar.

- These processes are reparented to PID 1 when the wrapper shell exits, so walking the terminal's process tree never saw them.
- Each scan reads the process table once and attributes remaining processes through inherited `SUPERSET_TERMINAL_ID` or `SUPERSET_PANE_ID`, propagating ownership from the nearest readable ancestor and leaving unreadable roots unattributed.
- Environment is re-read every scan instead of cached — a reused PID could otherwise be misattributed. Killed, aborted, or truncated inspections throw so partial output is never treated as a valid scan; on failure, previous state is preserved.
- Kill requests re-read the process table and re-check ownership and the listening socket before signalling, failing closed on any error or stale observation; macOS reads strip observed argv so arguments can't be mistaken for environment.
- Scope: the owning terminal must stay registered; attribution after it closes is not covered.

<sup>Written for commit 6ee0f6845e7390af0aec2a43ee4e0dc0e930d381. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/superset-sh/superset/pull/7256?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Port scanning now detects detached background processes and associates them with the correct terminal on Linux and macOS.
  * Process-scanning APIs support reusable process-table snapshots.
  * Terminal-owned servers can be safely stopped after current process and port ownership is verified.

* **Bug Fixes**
  * Improved handling of exited, inaccessible, or reused processes.
  * Prevented stale port ownership and controls after process replacement.
  * Added cancellation and timeout handling for process inspections.

* **Tests**
  * Expanded coverage for detached attribution, environment detection, filtering, cancellation, and stale-process scenarios.

### CODE REVIEW

File: packages/port-scanner/src/detached.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	}): Promise<Map<number, string>> {
+		const resolved = new Map<number, string>();
+		if (terminalIds.size === 0) return resolved;
+
+		const ppidByPid = new Map<number, number>();
+		for (const { pid, ppid } of table) ppidByPid.set(pid, ppid);
+
+		for (const [pid, entry] of this.cache) {
+			const ppid = ppidByPid.get(pid);
+			if (ppid === undefined || ppid !== entry.ppid) this.cache.delete(pid);
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/port-scanner/src/detached.ts]: Addressed in bdca4d9317. The current process table has no stable process-instance identifier, so the resolver now re-reads environments every scan instead of caching by PID/PPID. Added same-parent PID-reuse regressions (changed owner and no owner), a retry case for previously unreadable environments, and a PortManager regression confirming the old terminal no longer tracks or kills the replacement after rescanning. Tradeoff: environment reads measured 86–94 ms for 1,427 processes on this Mac; batching and scan cadence are unchanged. All 97 scanner tests pass.

File: packages/port-scanner/src/exec.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+			if (
+				execErr.name === "AbortError" ||
+				execErr.code === "ABORT_ERR" ||
+				execErr.killed ||
+				execErr.signal
+			) {
+				throw err;
+			}
+			if ("stdout" in execErr) {
+				return String(execErr.stdout ?? "");
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/port-scanner/src/exec.ts]: Addressed in bdca4d9317. stdout is now tolerated only when the error code is numeric, after the existing abort/kill checks. Real subprocess tests reproduce maxBuffer overflow and spawn failures being swallowed before the change and verify rejection afterward, while preserving stdout for numeric nonzero exits. All 97 scanner tests pass.

File: packages/port-scanner/src/procfs.ts [hunk 3]
```diff
@@ -290,3 +290,60 @@ export async function getListeningPortsLinuxProcfs(
...
+): Promise<Map<number, string | null>> {
+	const values = new Map<number, string | null>();
+	const limit = createLimiter(ENVIRON_READ_CONCURRENCY);
+	await Promise.all(
+		pids.map((pid) =>
+			limit(async () => {
+				signal?.throwIfAborted();
+				let content: string;
+				try {
+					content = await fs.readFile(`/proc/${pid}/environ`, "utf-8");
```
[the remark this answers is not in the document]
Comment (Kitenite) [reply packages/port-scanner/src/procfs.ts]: Addressed in bdca4d9317. Passed the scan signal into fs.readFile and rethrow read failures when the signal is aborted. A mocked-filesystem regression checks the forwarded signal and cancellation propagation; another verifies ordinary read failures still map to null. All 97 scanner tests and full typecheck pass.
